src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1506,28 +1510,92 @@ HashSet<string> GetPackageDependencies()
 
                 static string GetPackageId(LockFileTargetLibrary package) => $"{package.Name}/{package.Version.ToNormalizedString()}";
 
-                string GetPackageDiagnosticLevel(LockFileLibrary package)
+                Dictionary<string, LogLevel> GetPackageDiagnosticLevels()
```

Comment:
This method works by creating a dictionary representing the package graph. Then it can walk the graph by looking up the package id in the dictionary. When the number of log messages is much smaller than the number of packages, building the dictionary is inefficient, but walking the graph is efficient.

An alternative is to enumerate the packages list every time we need to find a package. But when we have many log messages, or if there's a deep path from a package with a warning to the direct package reference, then these lookups are very inefficient. Much much worse than the cost of creating the dictionary. I think the packages list is sorted, so the "enumeration" could actually be a binary search, but still it's not a dictionary lookup.

So, there's a design choice to make. Do we optimize for less memory (no dictionary) at the risk that some scenarios are much slower, or do we optimize for worst case at the cost that (probably) most cases will be a bit slower?  Although, given that MSBuild uses one process per parallel target execution, the cost of GC isn't as high (IMO) as a single process, multi-threaded app.

I don't think NuGet has sufficient telemetry to make a data driven decision, so we need to make a gut decision here.

Reply:
I would optimize for speed over memory... memory is less of a premium than speed in most cases and I would keep the general case fast. I think you did the right thing. Worst case, this will show up in a performance regression and if it's that bad, we can have that discussion for those specific scenarios.

Reply:
Kudos on the approach btw, I think a dictionary was the right choice and your code was easy to read.